bitmind/synthetic_image_generation/synthetic_image_generator.py

```diff
@@ -1,29 +1,201 @@
-        import tensorflow
-
-        warnings.filterwarnings("ignore", category=FutureWarning, module='diffusers')
-
-    def generate_image(self, annotation: dict, diffuser: DiffusionPipeline, save_dir: str):
-        """Generate images from annotations using a diffuser and save to the specified directory."""
-        with torch.no_grad():
-            prompt = annotation['description']
-            index = annotation.get('index', "missing_index")
-            generated_image = diffuser(prompt=prompt).images[0]
-            img = ToPILImage()(generated_image) if isinstance(generated_image, torch.Tensor) else generated_image
-            safe_prompt = prompt[:50].replace(' ', '_').replace('/', '_').replace('\\', '\\\\')
-            
-            img_filename = f"{save_dir}/{safe_prompt}-{index}.png"
-            return img, img_filename
\ No newline at end of file
+
+    def __init__(
+        self,
+        prompt_type='random',
+        prompt_generator_name=PROMPT_GENERATOR_NAMES[0],
+        diffuser_name=DIFFUSER_NAMES[0],
+        use_random_diffuser=False,
+        image_cache_dir=None
+    ):
+        assert prompt_type in PROMPT_TYPES,\
+            f'Invalid prompt type. Options are {prompt_type}'
+        assert prompt_generator_name in PROMPT_GENERATOR_NAMES,\
+            f'Invalid prompt generator name. Options are {PROMPT_GENERATOR_NAMES}'
+        assert use_random_diffuser or diffuser_name in DIFFUSER_NAMES, \
+            f'Invalid diffuser name. Options are {DIFFUSER_NAMES}'
+
+        self.use_random_diffuser = use_random_diffuser
+        self.prompt_type = prompt_type
+        self.prompt_generator_name = prompt_generator_name
+
+        if self.use_random_diffuser and diffuser_name is not None:
+            bt.logging.warning("Warning: diffuser_name will be ignored (use_random_diffuser=True)")
+            self.diffuser_name = None
+        else:
+            self.diffuser_name = diffuser_name
+
+        self.image_annotation_generator = None
+        if self.prompt_type == 'annotation':
+            bt.logging.info(f"Loading image captioning model ({IMAGE_ANNOTATION_MODEL})...")
+            self.image_annotation_generator = ImageAnnotationGenerator(model_name=IMAGE_ANNOTATION_MODEL)
+        else:
+            bt.logging.info(f"Loading prompt generation model ({prompt_generator_name})...")
+            self.prompt_generator = pipeline(
+                'text-generation', **PROMPT_GENERATOR_ARGS[prompt_generator_name])
+
+        self.image_cache_dir = image_cache_dir
+        if image_cache_dir is not None:
+            os.makedirs(self.image_cache_dir, exist_ok=True)
+
+        if diffuser_name is not None:
+            bt.logging.info(f"Loading image generation model ({diffuser_name})...")
+            self.diffuser = DiffusionPipeline.from_pretrained(
+                diffuser_name, torch_dtype=torch.float16, **DIFFUSER_ARGS[diffuser_name])
+            self.diffuser.to("cuda")
```

Comment:
i need to remove this, don't want to put this on the gpu just to clear it off and load the image captioning model